src/charm.py

```diff
@@ -118,6 +122,19 @@ def _on_generate_action(self, event: ActionEvent):
             "response": response["response"],
         })
 
+    def _on_pull_action(self, event: ActionEvent):
+        if not "model" in event.params:
+            event.fail("Parameter \"model\" was not provided, it is required.")
+            return
+        model = event.params["model"]
+
+        event.log(f"Downloading model {model}…")
+        requests.post(
+            f"http://localhost:{self._charm_state.port}/api/pull",
+            json={"name": model}
+        )
+        event.log("Done.")
+
     def _install_ollama(self):
         """ Download and install Ollama """
         run_shell("sudo snap install ollama")
```

Comment:
There's a [lib](https://github.com/canonical/operator-libs-linux/blob/main/lib/charms/operator_libs_linux/v2/snap.py) to manage snaps in vm charms. [Example](https://github.com/canonical/grafana-agent-operator/blob/main/src/snap_management.py).